md5sum.cpp

```diff
@@ -0,0 +1,217 @@
+
+        auto const space_pos = line.find(' ');
+        if(line.size() - 1 <= space_pos)
+            {
+            throw std::runtime_error(error_with_message("incorrect checksum line format"));
+            }
+
+        char const second_delimiter = line[space_pos + 1];
+        md5_file_mode file_mode;
+        if(second_delimiter == ' ')
+            {
+            file_mode = md5_file_mode::text;
+            }
+        else if(second_delimiter == '*')
+            {
+            file_mode = md5_file_mode::binary;
+            }
+        else
+            {
+            throw std::runtime_error(error_with_message("incorrect checksum line format"));
+            }
+
+        std::string md5sum = line.substr(0, space_pos);
+        std::string file = line.substr(space_pos + 2);
+
+        if(md5sum.size() != chars_per_formatted_hex_byte * md5len)
+            {
+            throw std::runtime_error(error_with_message("incorrect MD5 sum format"));
+            }
+
+        result_vec.emplace_back(std::move(file), std::move(md5sum), file_mode);
+        }
+
+    return result_vec;
+}
+
+std::vector<md5sum_for_file> md5_read_checksum_file(fs::path const& filename)
+{
+    auto const filename_string = filename.string();
+
+    std::ifstream is(filename_string);
+    if(!is)
+        {
+        std::ostringstream oss;
+        oss << "'" << filename_string << "': no such file or directory";
+        throw std::runtime_error(oss.str());
+        }
+
+    return md5_read_checksum_stream(is, filename_string);
+}
+
+std::string md5_calculate_stream_checksum
+    (std::istream& is
+    ,std::string const& stream_description
+    )
+{
+    std::vector<unsigned char> md5(md5len);
+
+    // Important: block_size must be a multiple of 64.
+    constexpr std::streamsize block_size = 4096;
+    md5_ctx ctx;
+    char buffer[block_size + 72];
+    std::streamsize read_count;
+
+    // Initialize the computation context.
+    md5_init_ctx(&ctx);
+
+    // Iterate over full file contents.
+    for(;;)
+        {
+        // We read the file in blocks of block_size bytes. One call of the
+        // computation function processes the whole buffer so that with the
+        // next round of the loop another block can be read.
+        is.read(buffer, block_size);
+        read_count = is.gcount();
+
+        // If end of file is reached, end the loop.
+        if (is.eof())
+            {
+            break;
+            }
+
+        if(read_count != block_size || !is)
+            {
+            std::ostringstream oss;
+            oss << "'" << stream_description << "': failed to compute md5sum";
+            throw std::runtime_error(oss.str());
+            }
+
+        // Process buffer with block_size bytes. Note that
+        // block_size % 64 == 0
+        md5_process_block(buffer, block_size, &ctx);
+        }
+
+    // Add the last bytes if necessary.
+    if(read_count > 0)
+        {
+        md5_process_bytes(buffer, read_count, &ctx);
```

Comment:
I'd add
```cpp
// Note that we have to use md5_process_bytes() and not the faster md5_process_block() here
// because the read_count is not necessarily a multiple of 64 here.
```